File: backend/src/main/java/com/sportstore/backend/service/StoreMapper.java
package com.sportstore.backend.service;

import com.sportstore.backend.domain.AppUser;
import com.sportstore.backend.domain.CartLine;
import com.sportstore.backend.domain.CustomerOrder;
import com.sportstore.backend.domain.OrderLine;
import com.sportstore.backend.domain.Product;
import com.sportstore.backend.dto.OrderDto;
import com.sportstore.backend.dto.ProductDto;
import com.sportstore.backend.dto.StoredCartDto;
import com.sportstore.backend.dto.StoredCartLineDto;
import com.sportstore.backend.dto.UserDto;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;
import org.springframework.stereotype.Component;

@Component
public class StoreMapper {

  public ProductDto toProductDto(Product product) {
    return new ProductDto(
      product.getId(),
      product.getName(),
      product.getCategory(),
      product.getDescription(),
      product.getPrice(),
      product.getImageUrl()
    );
  }

  public UserDto toUserDto(AppUser user) {
    return new UserDto(
      user.getId(),
      user.getRole().name().toLowerCase(),
      user.getNom(),
      user.getPrenom(),
      user.getAdresse(),
      user.getTelephone(),
      user.getUsername(),
      toStoredCartFromCartLines(user.getCartLines())
    );
  }

  public OrderDto toOrderDto(CustomerOrder order) {
    OrderDto dto = new OrderDto();
    dto.setId(order.getId());
    dto.setUserId(order.getUserId());
    dto.setUsername(order.getUsername());
    dto.setNom(order.getNom());
    dto.setPrenom(order.getPrenom());
    dto.setAdresse(order.getAdresse());
    dto.setTelephone(order.getTelephone());
    dto.setCreatedAt(order.getCreatedAt());
    dto.setTotal(order.getTotal());
    dto.setItemCount(order.getItemCount());
    dto.setShipped(order.isShipped());
    dto.setCart(toStoredCartFromOrderLines(order.getLines()));
    return dto;
  }

  public StoredCartDto toStoredCartFromCartLines(List<CartLine> lines) {
    List<StoredCartLineDto> dtos = new ArrayList<>();
    int itemCount = 0;
    BigDecimal total = BigDecimal.ZERO;

    for (CartLine line : lines) {
      dtos.add(new StoredCartLineDto(toProductDto(line.getProduct()), line.getQuantity()));
      itemCount += safeQuantity(line.getQuantity());
      total = total.add(line.getProduct().getPrice().multiply(BigDecimal.valueOf(safeQuantity(line.getQuantity()))));
    }

    return new StoredCartDto(dtos, itemCount, total);
  }

  public StoredCartDto toStoredCartFromOrderLines(List<OrderLine> lines) {
    List<StoredCartLineDto> dtos = new ArrayList<>();
    int itemCount = 0;
    BigDecimal total = BigDecimal.ZERO;

    for (OrderLine line : lines) {
      dtos.add(new StoredCartLineDto(toProductDto(line.getProduct()), line.getQuantity()));
      itemCount += safeQuantity(line.getQuantity());
      total = total.add(line.getProduct().getPrice().multiply(BigDecimal.valueOf(safeQuantity(line.getQuantity()))));
    }

    return new StoredCartDto(dtos, itemCount, total);
  }

  private int safeQuantity(Integer quantity) {
    return quantity == null ? 0 : Math.max(quantity, 0);
  }
}

File: backend-dotnet/Services/StoreMapper.cs
using SportStore.Api.Dtos;
using SportStore.Api.Models;

namespace SportStore.Api.Services;

public class StoreMapper
{
    public ProductDto ToProductDto(Product product) =>
        new(product.Id, product.Name, product.Category, product.Description, product.Price, product.ImageUrl);

    public UserDto ToUserDto(AppUser user) =>
        new(
            user.Id,
            user.Role.ToString().ToLowerInvariant(),
            user.Nom,
            user.Prenom,
            user.Adresse,
            user.Telephone,
            user.Username,
            ToStoredCartFromCartLines(user.CartLines)
        );

    public OrderDto ToOrderDto(CustomerOrder order) =>
        new()
        {
            Id = order.Id,
            UserId = order.UserId,
            Username = order.Username,
            Nom = order.Nom,
            Prenom = order.Prenom,
            Adresse = order.Adresse,
            Telephone = order.Telephone,
            CreatedAt = DateTime.SpecifyKind(order.CreatedAt, DateTimeKind.Utc),
            Total = order.Total,
            ItemCount = order.ItemCount,
            Shipped = order.Shipped,
            Cart = ToStoredCartFromOrderLines(order.Lines)
        };

    public StoredCartDto ToStoredCartFromCartLines(IEnumerable<CartLine> lines)
    {
        var list = lines
            .Select(line => new StoredCartLineDto(ToProductDto(line.Product), line.Quantity))
            .ToList();

        var itemCount = list.Sum(line => line.Quantity ?? 0);
        var total = list.Sum(line => (line.Product?.Price ?? 0m) * (line.Quantity ?? 0));
        return new StoredCartDto(list, itemCount, total);
    }

    public StoredCartDto ToStoredCartFromOrderLines(IEnumerable<OrderLine> lines)
    {
        var list = lines
            .Select(line => new StoredCartLineDto(ToProductDto(line.Product), line.Quantity))
            .ToList();

        var itemCount = list.Sum(line => line.Quantity ?? 0);
        var total = list.Sum(line => (line.Product?.Price ?? 0m) * (line.Quantity ?? 0));
        return new StoredCartDto(list, itemCount, total);
    }
}
